src/loki/polygon_search.cc

```diff
@@ -0,0 +1,115 @@
+#include "loki/polygon_search.h"
+
+namespace bg = boost::geometry;
+namespace vm = valhalla::midgard;
+namespace vb = valhalla::baldr;
+
+namespace valhalla {
+namespace loki {
+
+std::unordered_set<vb::GraphId> edges_in_rings(const std::vector<ring_bg_t>& rings,
+                                               baldr::GraphReader& reader,
+                                               const std::shared_ptr<sif::DynamicCost>& costing) {
+  // Get the lowest level and tiles
+  const auto tiles = vb::TileHierarchy::levels().back().tiles;
+  const auto bin_level = vb::TileHierarchy::levels().back().level;
+
+  // keep track which tile's bins intersect which rings
+  bins_collector bins_intersected;
+  std::unordered_set<vb::GraphId> avoid_edge_ids;
+
+  // first pull out all *unique* bins which intersect the rings
+  for (size_t ring_idx = 0; ring_idx < rings.size(); ring_idx++) {
+    auto ring = rings[ring_idx];
+    auto line_intersected = tiles.Intersect(ring);
+    for (const auto& tb : line_intersected) {
+      for (const auto& b : tb.second) {
+        bins_intersected[tb.first][b].push_back(ring_idx);
+      }
+    }
+  }
```

Comment:
what the comment says. we do need to record for each bin which ring**s** it intersects, as that could be multiple and that's important for intersecting the edges within later on.